### PR TITLE
implementation for bael-7192

### PR DESCRIPTION
**Pull request by imsurajmishra**
$SUBJECT

### CODE REVIEW

File: spring-boot-modules/spring-boot-logging-loki/src/test/java/com/baeldung/loki/DemoServiceLiveTest.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+@SpringBootTest(classes = DemoService.class)
+public class DemoServiceLiveTest {
+
+    @Test
+    public void givenLokiContainerRunning_whenDemoServiceInvoked_thenLokiAppenderCollectLogs() throws JsonProcessingException, InterruptedException {
+        DemoService service = new DemoService();
+        service.log();
+        Thread.sleep(2000);
```
Comment (theangrydev) [inline spring-boot-modules/spring-boot-logging-loki/src/test/java/com/baeldung/loki/DemoServiceLiveTest.java:38]: Why is this here? What condition are you waiting for?

We can use https://www.baeldung.com/awaitility-testing if there is a condition to wait on, rather than a busy sleep

File: spring-boot-modules/spring-boot-logging-loki/src/test/java/com/baeldung/loki/DemoServiceLiveTest.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+            JsonNode result = jsonNode.get("data").get("result").get(0).get("values");
+            result.iterator().forEachRemaining(e-> {
+                Iterator<JsonNode> elements = e.elements();
+                elements.forEachRemaining(f ->
+                    messages.add(f.toString())
+                );
+            });
+        } else {
+            System.out.println("Error: " + response.getStatusCodeValue());
+        }
```
Comment (theangrydev) [inline spring-boot-modules/spring-boot-logging-loki/src/test/java/com/baeldung/loki/DemoServiceLiveTest.java:82]: Let's instead make this an assertion

File: spring-boot-modules/spring-boot-logging-loki/src/test/java/com/baeldung/loki/DemoServiceLiveTest.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.client.RestTemplate;
+import org.springframework.web.util.UriComponentsBuilder;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+@SpringBootTest(classes = DemoService.class)
+public class DemoServiceLiveTest {
```
Comment (theangrydev) [inline spring-boot-modules/spring-boot-logging-loki/src/test/java/com/baeldung/loki/DemoServiceLiveTest.java:32]: Let's include briefly as a comment above the class what steps are required to run the Loki server externally

File: spring-boot-modules/spring-boot-logging-loki/src/test/java/com/baeldung/loki/DemoServiceLiveTest.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        HttpHeaders headers = new HttpHeaders();
+        headers.setContentType(MediaType.APPLICATION_JSON);
+        String query = "{level=\"INFO\"} |= `DemoService.log invoked`";
+
+        // Get current time in UTC
+        LocalDateTime currentDateTime = LocalDateTime.now(ZoneOffset.UTC);
+        String current_time_est = currentDateTime.format(DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss'Z'"));
+
+        LocalDateTime  tenMinsAgo = currentDateTime.minusMinutes(10);
+        String start_time_est = tenMinsAgo.format(DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss'Z'"));
```
Comment (theangrydev) [inline spring-boot-modules/spring-boot-logging-loki/src/test/java/com/baeldung/loki/DemoServiceLiveTest.java:51]: Do you mean UTC not EST?

File: spring-boot-modules/spring-boot-logging-loki/src/test/java/com/baeldung/loki/DemoServiceLiveTest.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+                elements.forEachRemaining(f ->
+                    messages.add(f.toString())
+                );
+            });
+        } else {
+            System.out.println("Error: " + response.getStatusCodeValue());
+        }
+
+        String expected = "DemoService.log invoked";
+        assertTrue(messages.stream().anyMatch(e -> e.contains(expected)));
```
Comment (theangrydev) [inline spring-boot-modules/spring-boot-logging-loki/src/test/java/com/baeldung/loki/DemoServiceLiveTest.java:85]:
```suggestion
        assertThat(messages).anyMatch(message -> message.contains(expected)));
```

File: spring-boot-modules/spring-boot-logging-loki/src/test/java/com/baeldung/loki/DemoServiceLiveTest.java [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+            String.class
+        );
+
+        List<String> messages = new ArrayList<>();
+        ObjectMapper objectMapper = new ObjectMapper();
+        if (response.getStatusCode() == HttpStatus.OK) {
+            String responseBody = response.getBody();
+            JsonNode jsonNode = objectMapper.readTree(responseBody);
+            JsonNode result = jsonNode.get("data").get("result").get(0).get("values");
+            result.iterator().forEachRemaining(e-> {
```
Comment (theangrydev) [inline spring-boot-modules/spring-boot-logging-loki/src/test/java/com/baeldung/loki/DemoServiceLiveTest.java:74]: Can you check the Baeldung formatter has been applied, the formatting here makes it look like it has not

**Review by theangrydev — CHANGES_REQUESTED** [review]

File: spring-boot-modules/spring-boot-logging-loki/src/test/java/com/baeldung/loki/DemoServiceLiveTest.java (hunk 1 above)
Comment (imsurajmishra) [reply spring-boot-modules/spring-boot-logging-loki/src/test/java/com/baeldung/loki/DemoServiceLiveTest.java]: removed!

File: spring-boot-modules/spring-boot-logging-loki/src/test/java/com/baeldung/loki/DemoServiceLiveTest.java (hunk 2 above)
Comment (imsurajmishra) [reply spring-boot-modules/spring-boot-logging-loki/src/test/java/com/baeldung/loki/DemoServiceLiveTest.java]: done , thanks.

File: spring-boot-modules/spring-boot-logging-loki/src/test/java/com/baeldung/loki/DemoServiceLiveTest.java (hunk 3 above)
Comment (imsurajmishra) [reply spring-boot-modules/spring-boot-logging-loki/src/test/java/com/baeldung/loki/DemoServiceLiveTest.java:31]: since we already describe in section #2, do we still need it here?

File: spring-boot-modules/spring-boot-logging-loki/src/test/java/com/baeldung/loki/DemoServiceLiveTest.java (hunk 4 above)
Comment (imsurajmishra) [reply spring-boot-modules/spring-boot-logging-loki/src/test/java/com/baeldung/loki/DemoServiceLiveTest.java]: ohh yes, thanks, changed it.

File: spring-boot-modules/spring-boot-logging-loki/src/test/java/com/baeldung/loki/DemoServiceLiveTest.java (hunk 6 above)
Comment (imsurajmishra) [reply spring-boot-modules/spring-boot-logging-loki/src/test/java/com/baeldung/loki/DemoServiceLiveTest.java]: done. thanks.

File: spring-boot-modules/spring-boot-logging-loki/src/test/java/com/baeldung/loki/DemoServiceLiveTest.java (hunk 3 above)
Comment (theangrydev) [reply spring-boot-modules/spring-boot-logging-loki/src/test/java/com/baeldung/loki/DemoServiceLiveTest.java:31]: I just mean add a comment like "this test assumes a running Loki server started using docker compose, for more details see the article"

File: spring-boot-modules/spring-boot-logging-loki/src/test/java/com/baeldung/loki/DemoServiceLiveTest.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+        HttpHeaders headers = new HttpHeaders();
+        headers.setContentType(MediaType.APPLICATION_JSON);
+        String query = "{level=\"INFO\"} |= `DemoService.log invoked`";
+
+        // Get current time in UTC
+        LocalDateTime currentDateTime = LocalDateTime.now(ZoneOffset.UTC);
+        String current_time_utc = currentDateTime.format(DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss'Z'"));
+
+        LocalDateTime tenMinsAgo = currentDateTime.minusMinutes(10);
+        String start_time_utc = tenMinsAgo.format(DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss'Z'"));
```
Comment (theangrydev) [inline spring-boot-modules/spring-boot-logging-loki/src/test/java/com/baeldung/loki/DemoServiceLiveTest.java:49]: camelCase for these please

**Review by theangrydev — CHANGES_REQUESTED** [review]

File: spring-boot-modules/spring-boot-logging-loki/src/test/java/com/baeldung/loki/DemoServiceLiveTest.java (hunk 7 above)
Comment (imsurajmishra) [reply spring-boot-modules/spring-boot-logging-loki/src/test/java/com/baeldung/loki/DemoServiceLiveTest.java]: done.

**Review by theangrydev — APPROVED** [review]

File: spring-boot-modules/spring-boot-logging-loki/src/main/java/com/baeldung/loki/DemoService.java [hunk 8]
```diff
@@ -0,0 +1,15 @@
+package com.baeldung.loki;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Service;
+
+@Service
+public class DemoService {
+
+    private final Logger LOG = LoggerFactory.getLogger(DemoService.class);
```
Comment (KevinGilmore) [inline spring-boot-modules/spring-boot-logging-loki/src/main/java/com/baeldung/loki/DemoService.java:10]: Conventionally, this should be a static final variable.
